tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,767 @@
+        SigningKeyConfig::ssm { .. } => Ok(HashMap::new()),
+    }
+}
+
+/// Creates keys using data stored in SigningKeyConfig enum
+/// Output: This method will has an in-place output meaning it edits key-id
+/// If key-id is populated, it will not change. Otherwise, it will be populated with a key-id of an available key
+async fn create_keys(signing_key_config: &mut SigningKeyConfig) -> Result<()> {
+    match signing_key_config {
+        SigningKeyConfig::file { path } => (),
+        SigningKeyConfig::kms { key_id, config, .. } => {
+            // QUESTION: keyid must remain at top level (we could encapsulate in config, but it would alter keyid reading for pubsys (which we alredy are technically doing by changing it to Option))
+            // TODO: don't override if key_id is not none 
+            let keyid_raw = config.create_kms_keys().await?;
+            *key_id = Some(keyid_raw);
+        } // QUESTION: unfortunately setting keyid for root key as well, but we don't use it
+        SigningKeyConfig::ssm { parameter } => (),
+    }
+    Ok(())
+}
+
+/// Must create a trait because can't directly implement a method for an struct in an external crate like KMSKeyConfig (which lives in pubsys-config/lib.rs)
+#[async_trait]
+trait KMSKeyConfigEx {
+    async fn create_kms_keys(&mut self) -> Result<String>;
+}
+
+/// Creates new KMS keys using cloudformation in regions specified
+/// Input Conditions: Alias+Region or AvailableKeys must be specified
+/// Output: Returns keyid and populates KMSKeyConfig with information about resources created
+/// available-keys starts as a map of pre-existing keyids:regions and will end as a map of pre-existing and generated keyids:regions
+/// key-stack-arns starts empty and will end as a map of keyids:stackarn if new keys are created
+#[async_trait]
+impl KMSKeyConfigEx for KMSKeyConfig {
+    async fn create_kms_keys(&mut self) -> Result<String> {
+        // Creating output params
+        let mut new_available_keys = HashMap::new();
+        let mut stack_arns = HashMap::new();
+
+        // TODO: if everything is null, do nothing (self.key_alias.is_none() && self.regions.is_some() && self.available_keys.is_none())
+        // Checking pre-conditions
+        if (self.key_alias.is_some() && self.regions.is_none())
+            || (self.key_alias.is_none() && self.regions.is_some())
+        {
+            error::KeyConfig.fail()?;
+        }
+        if self.available_keys.is_some() {
+            //available keys is either empty or populated with current values
+            new_available_keys = self
+                .available_keys
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: "available_keys",
+                })?
+                .clone();
+        }
+
+        // Generating new keys
+        // NOTE: using unwrap here because regions and alias existence is checked above
+        for region in self.regions.as_ref().unwrap().iter() {
```

Comment:
`regions` isn't guaranteed to always be `Some` so this would result in a panic. 
I think it would be best if we structured this method like so to deal with the different configuration combination:
```
match (self.regions, self.key_alias) {
    (None, None) => {
         // Both regions and key_alias are not specified
         // we don't need to create anything, so just no-op
    },
    (Some(regions), Some(key_alias)) => {
         // Both regions and key_alias are specified
         // we're good to go and create keys
    }
    (Some(regions), None) => {
         // Only regions is specified
         // probably error
    },
    (None, Some(key_alias)) => {
         // Only key_alias is specified
         // probably error
    },
}
```

This error cases can also just be combined under the catch-all case if you don't want to error differently.:
```
match (self.regions, self.key_alias) {
    (None, None) => {
         // Both regions and key_alias are not specified
         // we don't need to create anything, so just no-op
    },
    (Some(regions), Some(key_alias)) => {
         // Both regions and key_alias are specified
         // we're good to go
    },
    _ => {
         // error because configuration invalid.
    },
}
```